### ISSUE TITLE
BUG: fluid.layers.accuracy not work correctly

### ISSUE THREAD

**Issue by parap1uie-s** [PaddlePaddle/Paddle#21935]
现在accuracy的实现方法，是对input沿行方向取argmax后，取得最大概率下标，与稠密的label进行比对。

但当input的类别维度为1时，argmax结果显然恒为0。因此accuracy的取值将由label中0的占比来决定，而不是真正的“正确率”

可见[目前的实现](https://github.com/PaddlePaddle/Paddle/blob/release/1.6/python/paddle/fluid/layers/metric_op.py#L87):

```
def accuracy(input, label, k=1, correct=None, total=None):
    ......
    topk_out, topk_indices = nn.topk(input, k=k)  # Bug!!!
    acc_out = helper.create_variable_for_type_inference(dtype="float32")
```

如果输入的`input`是一个形如(bs, 1)的二分类模型输出时，显然topk输出结果恒为0，导致之后的准确率计算逻辑无意义

建议修复方案：

增加逻辑判断
```
def accuracy(input, label, k=1, correct=None, total=None):
    ......
    if input.shape[1] == 1:
        do something.
    else:
        topk_out, topk_indices = nn.topk(input, k=k)
    acc_out = helper.create_variable_for_type_inference(dtype="float32")
```

**Comment by parap1uie-s** [comment]
Duplicate of #18755

**Comment by baiyfbupt** [comment]
我理解`fliud.layers.accuracy`的输入input最后一维是m，就表示的是m分类问题
例如一个10分类问题的预测值predict应该是一个形状为`[batch_size, 10]`的Tensor
所以如果是2分类模型，是否可以让predict的形状为`[batch_size, 2]`而非`[batch_size, 1]`, 这样就可以正常计算了

**Comment by parap1uie-s** [comment]
> 我理解`fliud.layers.accuracy`的输入input最后一维是m，就表示的是m分类问题
> 例如一个10分类问题的预测值predict应该是一个形状为`[batch_size, 10]`的Tensor
> 所以如果是2分类模型，是否可以让predict的形状为`[batch_size, 2]`而非`[batch_size, 1]`, 这样就可以正常计算了

是的，设置二分类模型输出形状为(bs, 2)可以解决问题

但一方面，这依然是一个潜在风险，文档里也没有相关说明。
可以增加assert，要求输入tensor形状不为1

另一方面，理论上2维输出+softmax激活，因为最后fc层是交替获得梯度，模型收敛速度可能会有影响

详见：[link](https://stats.stackexchange.com/a/207067)

**Comment by baiyfbupt** [comment]
>但一方面，这依然是一个潜在风险，文档里也没有相关说明。
可以增加assert，要求输入tensor形状不为1

paddle这个accuracy接口默认是softmax_accuracy, 但是代码没有对输入的shape进行限制，文档也没有对input形状的数学意义进行解释，我也感觉这里可以改进一下，感谢反馈~

>另一方面，理论上2维输出+softmax激活，因为最后fc层是交替获得梯度，模型收敛速度可能会有影响

如果现在的用法可以满足需求可以先用，如果不能可以再提需求，paddle可能可以增加一个类似于TF的[BinaryAccuracy](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/BinaryAccuracy?hl=en&version=stable)解决问题
